File: Leetcode/Thek-thLexicographicalStringofAllHappyStringsofLengthn.rs
// https://leetcode.com/problems/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n

struct Solution {}

impl Solution {
    pub fn get_happy_string(n: i32, k: i32) -> String {
        let n = n as usize;
        let k = k as usize - 1;

        let l = [b'a', b'b', b'c'];
        let mut cur = vec![b'a'];

        for i in 1..n {
            for j in 0..3 {
                if cur[i - 1] != l[j] {
                    cur.push(l[j]);
                    break;
                }
            }
        }

        fn add(i: usize, cur: &mut Vec<u8>) -> bool {
            if i == 0 {
                if cur[0] == b'c' {
                    return false;
                } else {
                    cur[0] += 1;

                    return true;
                }
            } else {
                if cur[i] == b'c' {
                    if !add(i - 1, cur) {
                        return false;
                    }

                    if cur[i - 1] == b'a' {
                        cur[i] = b'b';
                    } else {
                        cur[i] = b'a';
                    }

                    return true;
                } else {
                    if cur[i] + 1 == cur[i - 1] {
                        if cur[i] == b'a' {
                            cur[i] += 2;
                        } else {
                            if !add(i - 1, cur) {
                                return false;
                            }

                            if cur[i - 1] == b'a' {
                                cur[i] = b'b';
                            } else {
                                cur[i] = b'a';
                            }
                        }
                    } else {
                        cur[i] += 1;
                    }

                    return true;
                }
            }
        }

        for _ in 0..k {
            if !add(n - 1, &mut cur) {
                return String::new();
            }
        }

        return String::from_utf8(cur).unwrap();
    }
}

File: Leetcode/Thek-thLexicographicalStringofAllHappyStringsofLengthn.cpp
// https://leetcode.com/problems/the-k-th-lexicographical-string-of-all-happy-strings-of-length-n

class Solution {
   public:
    string getHappyString(int n, int k) {
        auto check = [](const string& s) -> bool {
            for (size_t i = 0; i < s.size() - 1; ++i) {
                if (s[i] == s[i + 1]) {
                    return false;
                }
            }

            return true;
        };

        auto inc = [](string& s) -> bool {
            for (int i = s.length() - 1; i >= 0; --i) {
                if (s[i] < 'c') {
                    s[i] += 1;

                    return true;
                }
                s[i] = 'a';
            }

            return false;
        };

        string ans(n, 'a');
        ans[n - 1] = 'a' - 1;
        for (; k > 0; --k) {
            if (!inc(ans)) {
                return "";
            }
            while (!check(ans)) {
                if (!inc(ans)) {
                    return "";
                }
            }
        }

        return ans;
    }
};